core/src/main/java/org/elasticsearch/action/support/ActiveShardsWaiter.java

```diff
@@ -0,0 +1,145 @@
+     *                the input boolean indicates whether the operation timed out
+     * @param onUnacked the function to execute if the index creation cluster state update was not acknowledged;
+     *                  the input boolean indicates whether the operation timed out
+     * @return ActionListener for responding to index creation cluster state events and sending the action response
+     *         over the main listener, after waiting for the requested number of shards to be active
+     */
+    public ActionListener<ClusterStateUpdateResponse> wrapUpdateListenerWithWaiting(final CreateIndexClusterStateUpdateRequest request,
+                                                                                    final ActionListener<?> actionListener,
+                                                                                    final Consumer<Boolean> onAcked,
+                                                                                    final Consumer<Boolean> onUnacked) {
+        return new ActionListener<ClusterStateUpdateResponse>() {
+            @Override
+            public void onResponse(ClusterStateUpdateResponse response) {
+                if (response.isAcknowledged()) {
+                    final String indexName = request.index();
+                    final ActiveShardCount waitForActiveShards = request.waitForActiveShards();
+                    final TimeValue timeout = request.masterNodeTimeout();
+                    try {
+                        // the cluster state update with the created index has been acknowledged, now wait for the
+                        // configured number of active shards to be allocated before returning, as that is when indexing
+                        // operations can take place on the newly created index
+                        if (waitForActiveShards == ActiveShardCount.NONE) {
+                            // not waiting, so just run whatever we were to run when the waiting is
+                            onAcked.accept(false);
+                            return;
+                        }
+
+                        // wait for the configured number of active shards to be allocated before returning
+                        final ClusterStateObserver observer =
+                            new ClusterStateObserver(clusterService, logger, threadPool.getThreadContext());
+                        final ClusterStateObserver.ChangePredicate shardsAllocatedPredicate = new ClusterStateObserver.ChangePredicate() {
+                            @Override
+                            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                                return waitForActiveShards.enoughShardsActive(newState, indexName, settings);
```

Comment:
here (and two lines below) you pass the node settings and not the index settings. How did this pass any integration tests? I would also prefer to change the settings parameter of this method to indexmetadata to avoid these confusions.